[Lee Zhi Xuan] iP

leezhixuan opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” –  David Allen ([source](https://dansilvestre.com/productivity-quotes))

![Image of Shin-chan](https://p.kindpng.com/picc/s/490-4904208_cartoon-japan-and-shinchan-image-shinchan-cartoon-images.png)

DukePro frees your mind of having to remember things you need to do. It's,

 - text-based
 - easy to learn
 - ~~FAST~~ SUPER FAST to use

All you need to do is,

1. Download it from here.
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉

Most importantly, it is **FREE**!

Features:

 - [x] Managing tasks
 - [ ] Managing deadlines (coming soon)
 - [ ] Reminders (coming soon)

---
If you Java programmer, you can use it to practice Java too. Here's the main method:
```java
public class Launcher {

    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

leezhixuan committed:
Complete Level-1

leezhixuan committed:
Complete Level-2: Add, List

leezhixuan committed:
Complete Level-3: Mark as Done

leezhixuan committed:
Complete Level-4: ToDos, Events, Deadlines

leezhixuan committed:
Complete A-TextUiTesting

leezhixuan committed:
Complete Level 5: Handle Errors

leezhixuan committed:
Complete A-Exceptions

leezhixuan committed:
Complete Level 6: Delete

leezhixuan committed:
Complete A-Enums

leezhixuan committed:
Complete Level-7

leezhixuan committed:
Add data dDirectory

leezhixuan committed:
Complete Level-8

leezhixuan committed:
Merge branch-Level-7 and branch-Level-8 into master

leezhixuan committed:
Add Storage and Ui Class

leezhixuan committed:
Add Parser

leezhixuan committed:
Update Enum

leezhixuan committed:
Complete A-MoreOOP

leezhixuan committed:
Complete A-Packages

leezhixuan committed:
Complete A-JUnit

leezhixuan committed:
Add JavaDoc

leezhixuan committed:
Meet Coding Standard

leezhixuan committed:
Complete Find

leezhixuan committed:
Merge branch 'branch-Level-9'

leezhixuan committed:
Merge branch-A-CodingStandard in Level-9

leezhixuan committed:
Merge branch-A-JavaDocs in Level-9

leezhixuan committed:
Merge add-gradle-support into branch-A-Gradle

leezhixuan committed:
Complete A-Gradle and A-Checkstyle

cookiedan42 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,204 @@
+package duke;
+
+import duke.command.*;
```
Imported classes should always be listed explicitly.

leezhixuan replied:
Thanks for pointing it out :)

cookiedan42 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,204 @@
+            } catch (DukeException e) {
+                ui.printProper(e.getMessage());
+            }
+        } else if (command.startsWith(Commands.EVENT.asLowerCase())) {
```
Logical units within a block should be separated by one blank line.
Consider placing a new line between  each of the command.startsWith() blocks to clearly divide the sections for dealing with different commands

eezj35 replied:
Yup agreed, a blank line to separate each if else block will help with readability!

leezhixuan replied:
Thanks for pointing it out, I'll fix it in the next push

cookiedan42 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,204 @@
+            } else if (command.substring(5).isBlank()) {
+                throw new DukeException("Really? An event of nothing?");
+            }
+        } else {
```
Logical units within a block should be separated by one blank line.
Consider placing a new line between  each of the command.startsWith() blocks to clearly divide the sections for dealing with different commands

cookiedan42 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,155 @@
+    protected static void createTaskListStorage() {
+        //creates a file to store content
+        File s = new File("./data");
+        boolean sResult;
```
Boolean variables/methods should be named to sound like booleans

sResult and fResult

eezj35 replied:
Yup agreed

leezhixuan replied:
My bad, fixed this!

cookiedan42 commented on src/main/java/duke/ToDoList.java:
```diff
@@ -0,0 +1,171 @@
+        return this.record.get(index);
+    }
+
+    public void find(String target) {
```
Write descriptive header comments for all public classes/methods.

cookiedan42 reviewed (commented):
LTGM

zhuoyang125 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,204 @@
+     * @param storage The Storage that the chat bot uses.
+     * @return The type of Command to be executed.
+     */
+    protected Command parse(String command, ToDoList tdl, Ui ui, Duke chatBot, Storage storage) {
```
This looks to be a pretty long method 🤔. It might be better to add comments at certain branches of the code so that it's clear what's going on when different conditions are met.

eezj35 replied:
Perhaps you could split the if else blocks into separate functions instead?

leezhixuan replied:
Thanks for the tips, I think I will split them up :)

zhuoyang125 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,204 @@
+                formatChecker(command);
+                String substring = command.substring(5);
+                return new ToDoCommand(tdl, substring);
+            } catch (StringIndexOutOfBoundsException e) {
```
It seems like similar conditions are being caught at each branch condition. Perhaps consider abstracting exception handling and printing into another function, this may minimize code duplication!

zhuoyang125 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,155 @@
+        //creates a file to store content
+        File s = new File("./data");
+        boolean sResult;
+        boolean fResult;
```
It is also not very clear what fResult and sResult refer to... Perhaps having clearer names will help others to understand this method.

eezj35 replied:
Yes agreed, it took me quite awhile to understand what each of these booleans are supposed to be for

zhuoyang125 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,155 @@
+                    break;
+                }
+                char type = str.charAt(3);
+                char status = str.charAt(6);
```
Numbers like '3' and '6' in the code seem like magic numbers, as their meanings are not explicitly stated. Perhaps it would be good to store these numbers in variables or add comments to explain why the char variables refer to type and status.

leezhixuan replied:
Got it, I've included an explanation. Thanks for letting me know :)

zhuoyang125 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,155 @@
+                    try {
+                        String temp = str.substring(9);
+                        String item = temp.substring(0, temp.indexOf(" ")); //name
+                        String temp2 = temp.substring(temp.indexOf("("));
```
Perhaps 'temp' and 'temp2' can be renamed to something that is clearer?

zhuoyang125 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,155 @@
+                    }
+                } else {
+                    if (a instanceof Deadline) {
+                        fw.write(number.toString() + "." + a.logo() + "[ ] " + a.getName()
```
The code to format the tasks when they are written to file seems to be duplicated too. Perhaps you could consider abstracting the file formatting to another method?

zhuoyang125 commented on src/main/java/duke/ToDoList.java:
```diff
@@ -0,0 +1,171 @@
+        System.out.println("========== " + this.name + " ===========");
+        System.out.println("Good! Good! Took you long enough to complete this:");
+        System.out.println("  " + "[X] " + record.get(num-1).toString());
+        System.out.println("========== " + this.name + " ===========\n");
```
Perhaps it would be better for the `println` statements to be handled by the Ui class?

eezj35 replied:
Yup agreed, it will make your code slightly more readable as well!

leezhixuan replied:
Wow, hadn't thought of it like this. I deleted the entire Ui class lmao.... :/

zhuoyang125 commented on src/main/java/duke/ToDoList.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    public void find(String target) {
+        ArrayList<Task> temp = new ArrayList<>();
```
Perhaps rename `temp` to something more descriptive such as `resultFromFind`? Especially since `find` is a rather long method and `temp` is used later on.

leezhixuan replied:
Good point, resolved it!

zhuoyang125 reviewed (commented):
I feel that the code is structured quite well and there is good use of OOP! Maybe consider renaming certain methods/variables and reduce the use of duplicate code to make the code base easier to understand.

eezj35 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,58 @@
+
+    private static String name = "Duke";
+    private boolean isRunning = false;
+    private ToDoList tdl;
```
Maybe for the tdl variable, you could type change to toDoList? so it would be easier to read your code later on 👍

eezj35 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    private void startBot() {
+        this.isRunning = true;
+        Scanner input = new Scanner(System.in);
+        while (this.isRunning) {
+            String command = input.nextLine();
+            Command c = this.parser.parse(command, this.tdl, this.ui, this, this.storage);
+            c.execute();
```
I like how you wrote your code here! Very easy to read!

leezhixuan replied:
Many thanks!

eezj35 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,204 @@
+            @Override
+            public String asLowerCase() {
+                return LIST.toString().toLowerCase();
+            }
+        },
+        DONE {
+            @Override
+            public String asLowerCase() {
+                return DONE.toString().toLowerCase();
+            }
+        },
+        TODO{
+            @Override
+            public String asLowerCase() {
+                return TODO.toString().toLowerCase();
+            }
+        },
+        EVENT{
+            @Override
+            public String asLowerCase() {
+                return EVENT.toString().toLowerCase();
+            }
+        },
+        DEADLINE{
+            @Override
+            public String asLowerCase() {
+                return DEADLINE.toString().toLowerCase();
+            }
+        },
+        DELETE{
+            @Override
+            public String asLowerCase() {
+                return DELETE.toString().toLowerCase();
+            }
+        },
+        FIND{
+            @Override
+            public String asLowerCase() {
+                return FIND.toString().toLowerCase();
+            }
```
I like your use of enums for this!

leezhixuan replied:
thank you very much :))

eezj35 reviewed (commented):
Great job following the coding standards! Nothing much to comment on

leezhixuan committed:
Complete Level-10

leezhixuan committed:
Refactor code

leezhixuan committed:
Fix issues raised in PR review

leezhixuan committed:
Complete A-Assertions

Assertions are used to define assumptions about the program state so that the runtime can verify them. An assertion failure indicates a possible bug in the code because the code has resulted in a program state that violates an assumption about how the code should behave.

I have used assertions to ensure that a Command is always returned after an input of any type is parsed.

leezhixuan committed:
Complete A-CodeQuality

Among various dimensions of code quality, such as run-time efficiency, security, and robustness, one of the most important is understandability. High code quality means that the code can be understood with ease.

I have added an abstract method to my enum class that is responsible for returning the correct offset value. This means that "magic numbers" throughout the Parser class can be removed. I have also initialised variables containing the offset value in the Storage class for the same purpose.

leezhixuan committed:
Complete A-Streams

Streams are more expressive and readable than traditional for-loops.

I refactored my FindCommand and the find method in ToDoList to incorporate the use of streams. This greatly reduced the length of the code and the need for for-loops.

leezhixuan committed:
Merge pull request #1 from leezhixuan/branch-A-Streams

Complete A-Streams

leezhixuan committed:
Add MANIFEST.MF

leezhixuan committed:
Merge branch 'master' into branch-A-Assertions

leezhixuan committed:
Merge pull request #2 from leezhixuan/branch-A-Assertions

Branch A-Assertions

leezhixuan committed:
Merge branch 'master' into branch-A-CodeQuality

leezhixuan committed:
Merge pull request #3 from leezhixuan/branch-A-CodeQuality

Branch A-CodeQuality

leezhixuan committed:
Create gradle.yml

leezhixuan committed:
Fix checkstyle issues

leezhixuan committed:
Implement C-BetterSearch

leezhixuan committed:
Complete A-BetterGui

leezhixuan committed:
Update gradle.yml

Removed I/O redirection test for Linux, MacOS and Windows

leezhixuan committed:
Minor tweaks to GUI

leezhixuan committed:
Fix spacings

leezhixuan committed:
Add Ui.png

leezhixuan committed:
Merge branch-A-BetterGui with master

leezhixuan committed:
Set theme jekyll-theme-midnight

leezhixuan committed:
Set theme jekyll-theme-minimal

leezhixuan committed:
Set theme jekyll-theme-architect

leezhixuan committed:
Minor changes to README.md

leezhixuan committed:
Merge branch 'master' of github.com:leezhixuan/ip

leezhixuan committed:
Improve code quality by applying SLAP
